crypto/sha/asm/keccak1600-avx512.pl

```diff
@@ -0,0 +1,449 @@
+# suitable for Theta with intra-register permutations alone. Here is
+# "magic" layout for Chi (with pre-Theta shuffle):
+#
+# [4][4] [3][3] [2][2] [1][1] [0][0]>4.3.2.1.0>[4][4] [3][3] [2][2] [1][1] [0][0]
+# [4][0] [3][4] [2][3] [1][2] [0][1]>3.2.1.0.4>[3][4] [2][3] [1][2] [0][1] [4][0]
+# [4][1] [3][0] [2][4] [1][3] [0][2]>2.1.0.4.3>[2][4] [1][3] [0][2] [4][1] [3][0]
+# [4][2] [3][1] [2][0] [1][4] [0][3]>1.0.4.3.2>[1][4] [0][3] [4][2] [3][1] [2][0]
+# [4][3] [3][2] [2][1] [1][0] [0][4]>0.4.3.2.1>[0][4] [4][3] [3][2] [2][1] [1][0]
+#
+# Layout suitable to Theta has x coordinates aligned column-wise
+# [it's interleaved with Pi indices transformation for reference]:
+#
+# [4][4] [3][3] [2][2] [1][1] [0][0]	$A00
+##[0][4] [0][3] [0][2] [0][1] [0][0]
+# [3][4] [2][3] [1][2] [0][1] [4][0]	$A01
+##[2][3] [2][2] [2][1] [2][0] [2][4]
+# [2][4] [1][3] [0][2] [4][1] [3][0]	$A02
+##[4][2] [4][1] [4][0] [4][4] [4][3]
+# [1][4] [0][3] [4][2] [3][1] [2][0]	$A03
+##[1][1] [1][0] [1][4] [1][3] [1][2]
+# [0][4] [4][3] [3][2] [2][1] [1][0]	$A04
+##[3][0] [3][4] [3][3] [3][2] [3][1]
+#
+# Pi itself is performed by blending above data and finally shuffling it
+# to original Chi layout:
+#
+# [1][1] [2][2] [3][3] [4][4] [0][0]>1.2.3.4.0>[4][4] [3][3] [2][2] [1][1] [0][0]
+# [2][3] [3][4] [4][0] [0][1] [1][2]>2.3.4.0.1>[4][0] [3][4] [2][3] [1][2] [0][1]
+# [3][0] [4][1] [0][2] [1][3] [2][4]>3.4.0.1.2>[4][1] [3][0] [2][4] [1][3] [0][2]
+# [4][2] [0][3] [1][4] [2][0] [3][1]>4.0.1.2.3>[4][2] [3][1] [2][0] [1][4] [0][3]
+# [0][4] [1][0] [2][1] [3][2] [4][3]>0.1.2.3.4>[4][3] [3][2] [2][1] [1][0] [0][4]
+#
+# As implied, data is loaded in Chi layout. Digits in variables' names
+# represent right most coordinates of loaded data chunk:
+
+my ($A00,	# [4][4] [3][3] [2][2] [1][1] [0][0]
+    $A01,	# [4][0] [3][4] [2][3] [1][2] [0][1]
+    $A02,	# [4][1] [3][0] [2][4] [1][3] [0][2]
+    $A03,	# [4][2] [3][1] [2][0] [1][4] [0][3]
+    $A04) =	# [4][3] [3][2] [2][1] [1][0] [0][4]
+    map("%zmm$_",(0..4));
+
+# We also need to map the magic order into offsets within structure:
+
+my @A_jagged = ([0,0], [1,0], [2,0], [3,0], [4,0],
+		[4,1], [0,1], [1,1], [2,1], [3,1],
+		[3,2], [4,2], [0,2], [1,2], [2,2],
+		[2,3], [3,3], [4,3], [0,3], [1,3],
+		[1,4], [2,4], [3,4], [4,4], [0,4]);
+   @A_jagged_in  = map(8*($$_[0]*8+$$_[1]), @A_jagged);	# ... and now linear
+   @A_jagged_out = map(8*($$_[0]*5+$$_[1]), @A_jagged);	# ... and now linear
+
+my @T       = map("%zmm$_",(5..7,16..17));
+my @Chi     = map("%zmm$_",(18..22));
+my @Theta   = map("%zmm$_",(33,23..26));
+my @Rhotate = map("%zmm$_",(27..31));
+
+my ($C00,$D00) = @T[0..1];
+my ($k00001,$k00010,$k00100,$k01000,$k10000,$k11111) = map("%k$_",(1..6));
+
+$code.=<<___;
+.text
+
+.type	__KeccakF1600,\@function
+.align	32
+__KeccakF1600:
+	lea		iotas(%rip),%r10
+	mov		\$24,%eax
+	jmp		.Loop_avx512
+
+.align	32
+.Loop_avx512:
+	######################################### Theta
+	#vpermq		$A00,@Theta[0],$A00
```

Comment:
Humans tend to appreciate symmetry. Originally code was written as "we start by re-arranging all elements of input vectors". But truth is that $A00 is order that doesn't need rearrangement. Even though it was apparent all along it was added to satisfy the human urge for symmetry. And now it's kept as commentary, to satisfy the drift :-)